src/napari_spatialdata/_sdata_widgets.py

```diff
@@ -97,46 +118,55 @@ def _update_layers_visibility(self) -> None:
         """Toggle layer visibility dependent on presence in currently selected coordinate system."""
         elements = self.elements_widget._elements
         coordinate_system = self.coordinate_system_widget._system
-
         # No layer selected on first time coordinate system selection
         if self.viewer_model.viewer.layers:
             for layer in self.viewer_model.viewer.layers:
-                if layer.name not in elements:
-                    layer.visible = False
-                elif layer.metadata["_active_in_cs"]:
-                    layer.visible = True
-                    # Prevent _update_visible_in_coordinate_system of invalid removal of coordinate system
-                    layer.metadata["_active_in_cs"].add(coordinate_system)
-                    layer.metadata["_current_cs"] = coordinate_system
-
-    def _add_circles(self, key: str) -> None:
+                if "sdata" in layer.metadata:
+                    if layer.name not in elements:
+                        layer.visible = False
+                    elif layer.metadata["_active_in_cs"]:
+                        layer.visible = True
+                        # Prevent _update_visible_in_coordinate_system of invalid removal of coordinate system
+                        layer.metadata["_active_in_cs"].add(coordinate_system)
+                        layer.metadata["_current_cs"] = coordinate_system
+
+    def _add_circles(self, sdata: SpatialData, key: str, multi: bool) -> None:
         selected_cs = self.coordinate_system_widget._system
-        self.viewer_model.add_sdata_circles(self._sdata, selected_cs, key)
+        self.viewer_model.add_sdata_circles(sdata, selected_cs, key, multi)
 
-    def _add_polygons(self, key: str) -> None:
+    def _add_polygons(self, sdata: SpatialData, key: str, multi: bool) -> None:
         selected_cs = self.coordinate_system_widget._system
-        self.viewer_model.add_sdata_shapes(self._sdata, selected_cs, key)
+        self.viewer_model.add_sdata_shapes(sdata, selected_cs, key, multi)
 
     def _add_shapes(self, key: str) -> None:
-        if type(self._sdata.shapes[key].iloc[0][0]) == shapely.geometry.point.Point:
-            self._add_circles(key)
-        elif (type(self._sdata.shapes[key].iloc[0][0]) == shapely.geometry.polygon.Polygon) or (
-            type(self._sdata.shapes[key].iloc[0][0]) == shapely.geometry.multipolygon.MultiPolygon
+        sdata, multi = _get_sdata_key(self._sdata, self.elements_widget._elements, key)
+
+        original_name = key
+        if multi:
+            original_name = original_name[: original_name.rfind("_")]
```

Comment:
I'd add some validations here, like that what is at the right side of _ is a number. This to recognize cases like this:

1. The user adds two spatialdata objects
2. The user removes some layers, in particular "my_layer_0"
3. The user manually renames "my_layer_1" to "my_layer".
4. Here original name would lead to "my", leading to a wrong behavior

Reply:
This function is only called when adding the layer. Given that the `element_name` is created in `SdataWidget` this should hold true still. A user
can only change the layer name at the moment, but this does not change the element name. Only when saving a 
renamed layer back do we need to perform extra checks.

Reply:
> A user can only change the layer name at the moment, but this does not change the element name. 

why do we want the user to rename a layer? I would really discourage that.

Reply:
> why do we want the user to rename a layer? I would really discourage that.

As Wouter said this could happen due to the napari design, but actually reading the previous Wouter's comment, it should not be a problem since the new name would not be used internally. So I think that this issue can be resolved.